### PR TITLE
PSM Data Feed Addition

### PR DESCRIPTION
**Pull request by Michael7371**
# PR Details
## Description

This PR introduces support for PSM message processing to the ODE. Support for both required and optional fields is present, however I do not have a sample message that covers any of the optional fields for testing. All required fields have been tested with sample PSM messages generated from the NJDOT deployment.

This pipeline takes in PSM messages on the new UDP port (44940) and outputs processed JSON PSM messages on the topic: topic.OdePsmJson. As with other data feeds, multiple other topics are used for internal processing/management: topic.OdeRawEncodedPSMJson and topic.OdePsmTxPojo.

## Related Issue

There are no open GitHub issues related to this PR.

## Motivation and Context

Required for processing Personal Safety Messages through the ODE. 

## How Has This Been Tested?

I followed the convention from previous data feeds for unit testing and verified message processing with sample messages from the NJDOT deployment.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ x ] I have added any new packages to the sonar-scanner.properties file
- [ x ] My change requires a change to the documentation.
- [ x ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [ x ] I have added tests to cover my changes.
- [ x ] All new and existing tests passed.

### CODE REVIEW

**Review by drewjj — CHANGES_REQUESTED** [review]
Everything checks out, runs well and tests well. I think this is ready to go but do you mind editing this schema documentation to include PSM? Link to the schema you have like the others do. It is an easier, more human, way to find the schemas.

**Comment by Michael7371** [comment]
> Everything checks out, runs well and tests well. I think this is ready to go but do you mind editing this schema documentation to include PSM? Link to the schema you have like the others do. It is an easier, more human, way to find the schemas.
> 
> https://github.com/Michael7371/jpo-ode/tree/psm-data-feed/docs/schemas

Gotcha, that should be updated now.

**Review by drewjj — APPROVED** [review]
Great, Looks good!

**Review by dmccoystephenson — CHANGES_REQUESTED** [review]
With the addition of a new message type, the following should be done:
- The [ODE Data Flow Overview diagram](https://github.com/CDOT-CV/jpo-ode/blob/master/docs/data-flow-diagrams/ODE%20Data%20Flow%20Overview.drawio.png) should be updated to include the path that PSMs take.
- The existing overview data flow explanations should be updated as necessary in the README in the `docs/data-flow-diagrams` directory.
- A new subdirectory 'psm' should be added to the `docs/data-flow-diagrams` directory.
- In that new subdirectory a new diagram for the data flow of PSMs should be created.
- A new data flow explanation for PSMs should be added to the README in the `docs/data-flow-diagrams` directory.

**Comment by Michael7371** [comment]
@dmccoystephenson I pushed an update that addresses most of your comments. I didn't see what needed to be added to the ODE Data Flow Diagram, so I left that unchanged.

File: jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package us.dot.its.jpo.ode.model;
+
+
+public class OdePsmMetadata extends OdeLogMetadata {
+
+	/**
+	 *
```
Comment (dmccoystephenson) [inline jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:7]: This comment seems unnecessary

File: jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		super();
+	}
+
+	public OdePsmMetadata(OdeMsgPayload payload) {
+		super(payload);
+	}
+
+	public OdePsmMetadata(OdeMsgPayload payload, SerialId serialId, String receivedAt) {
+
+	}
```
Comment (dmccoystephenson) [inline jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:30]: Is this constructor necessary?

File: jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmPayload.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.fasterxml.jackson.annotation.*;
+
+import us.dot.its.jpo.ode.plugin.j2735.J2735PSM;
+
+public class OdePsmPayload extends OdeMsgPayload {
+
+	   /**
+	 * 
+	 */
```
Comment (dmccoystephenson) [inline jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmPayload.java:11]: Same here, this comment can probably be removed

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PSM.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
+package us.dot.its.jpo.ode.plugin.j2735;
+
+import us.dot.its.jpo.ode.plugin.asn1.Asn1Object;
+
+public class J2735PSM extends Asn1Object {
+
+	/**
+	 * 
+	 */
```
Comment (dmccoystephenson) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PSM.java:9]: This comment should be removed or populated

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package us.dot.its.jpo.ode.plugin.j2735;
+
+public enum J2735PersonalDeviceUsageState {
+    unavailable,
+    aPEDESTRIAN,
+    aPEDALCYCLIST,
+    aPUBLICSAFETYWORKER,
+    anANIMAL
```
Comment (dmccoystephenson) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java:8]: Are these the correct values for this enum? Looks like the same ones as the J2735PersonalDeviceUserType enum.

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PropelledInformation.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package us.dot.its.jpo.ode.plugin.j2735;
+
+import us.dot.its.jpo.ode.plugin.asn1.Asn1Object;
+
+public class J2735PropelledInformation extends Asn1Object  {
+	
+    /**
+	 * 
+	 */
```
Comment (dmccoystephenson) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PropelledInformation.java:9]: This comment can be removed

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/OdePsmDataCreatorHelper.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import us.dot.its.jpo.ode.model.RxSource;
+import us.dot.its.jpo.ode.plugin.j2735.builders.PSMBuilder;
+import us.dot.its.jpo.ode.util.JsonUtils;
+import us.dot.its.jpo.ode.util.XmlUtils;
+import us.dot.its.jpo.ode.util.XmlUtils.XmlUtilsException;
+
+public class OdePsmDataCreatorHelper {
+
+	public OdePsmDataCreatorHelper() {
+	}
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/OdePsmDataCreatorHelper.java:24]: Is this constructor necessary?

**Review by dmccoystephenson — CHANGES_REQUESTED** [review]
Looks good, just a few comments!

Also, the architecture diagram should be updated to include PSMs as input:
https://github.com/CDOT-CV/jpo-ode/blob/master/docs/images/readme/figure1.png

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java (hunk 5 above)
Comment (Michael7371) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java:8]: Good catch yeah I'll fix that.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/OdePsmDataCreatorHelper.java (hunk 7 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/OdePsmDataCreatorHelper.java:24]: Honestly, I am not too sure. I just followed the convention with all the other Ode{dataflow name}DataCreatorHelper.java files.

File: jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java (hunk 2 above)
Comment (Michael7371) [reply jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:30]: No, just came over from the BSM implementation (not sure it's useful over there either).

Comment (Michael7371) [reply jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:30]: I'll remove it

**Comment by Michael7371** [comment]
> Looks good, just a few comments! All tests are passing.
> 
> One thing: The architecture diagram should be updated to include PSMs as input: https://github.com/CDOT-CV/jpo-ode/blob/master/docs/images/readme/figure1.png

Do you know where the source file that generated that figure is stored?

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java (hunk 5 above)
Comment (dmccoystephenson) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java]: Thanks!

**Review by dmccoystephenson — APPROVED** [review]
This looks good to me!

**Pull request merged**